### PR TITLE
Refactor SKR watcher installation to use native Kubernetes client

### PR DESCRIPTION
**Pull request by khlifi411**
implementation of #301

### CODE REVIEW

**Comment by adityabhatia** [comment]
Updated module-manager lib with the latest commit

File: controllers/kyma_controller.go [hunk 1]
```diff
@@ -202,23 +203,25 @@ func (r *KymaReconciler) syncModuleCatalog(ctx context.Context, syncContext *rem
...
+func (r *KymaReconciler) stateHandling(ctx context.Context, kyma *v1alpha1.Kyma,
+	syncContext *remote.KymaSynchronizationContext,
+) (ctrl.Result, error) {
 	log.FromContext(ctx).Info("syncing state", "state", string(kyma.Status.State))
 	switch kyma.Status.State {
 	case "":
 		return ctrl.Result{}, r.HandleInitialState(ctx, kyma)
 	case v1alpha1.StateProcessing:
-		return ctrl.Result{Requeue: true}, r.HandleProcessingState(ctx, kyma)
+		return ctrl.Result{Requeue: true}, r.HandleProcessingState(ctx, kyma, syncContext)
```
[the remark this answers is not in the document]
Comment (adityabhatia) [reply controllers/kyma_controller.go:214]: The `ctx` is set during controller start, so that is not a good option, even if use unique kyma keys as keys for `context.WithValue(...)`

File: pkg/deploy/chart_manager.go [hunk 2]
```diff
@@ -6,26 +6,28 @@ import (
...
+
 var ErrExpectedExactlyOneSKRConfig = errors.New("expected exactly one SKR config")
 
 type SKRWebhookChartManager interface {
-	InstallWebhookChart(ctx context.Context, kyma *v1alpha1.Kyma,
-		remoteClientCache *remote.ClientCache, kcpClient client.Client) (bool, error)
-	RemoveWebhookChart(ctx context.Context, kyma *v1alpha1.Kyma,
-		syncCtx *remote.KymaSynchronizationContext) error
+	InstallWebhookChart(ctx context.Context, syncContext *remote.KymaSynchronizationContext) (bool, error)
+	RemoveWebhookChart(ctx context.Context, syncContext *remote.KymaSynchronizationContext) error
```
Comment (adityabhatia) [inline pkg/deploy/chart_manager.go:30]: the file is called `chart_manager.go`, with a `SKRWebhookChartManager`, with method names `InstallWebhookChart`, too much duplicacy. Could we come up with better names?

File: pkg/deploy/chart_manager.go [hunk 3]
```diff
@@ -74,81 +76,97 @@ type SkrChartConfig struct {
...
-func (m *EnabledSKRWebhookChartManager) InstallWebhookChart(ctx context.Context, kyma *v1alpha1.Kyma,
-	remoteClientCache *remote.ClientCache, kcpClient client.Client,
+func (m *EnabledSKRWebhookChartManager) InstallWebhookChart(ctx context.Context,
+	syncContext *remote.KymaSynchronizationContext,
 ) (bool, error) {
-	skrClient, err := remote.NewRemoteClient(ctx, kcpClient, client.ObjectKeyFromObject(kyma),
-		kyma.Spec.Sync.Strategy, remoteClientCache)
+	logger := logf.FromContext(ctx)
+	kymaObjKey := client.ObjectKeyFromObject(syncContext.ControlPlaneKyma)
+	skrCfg, err := remote.GetRemoteRestConfig(ctx, syncContext.ControlPlaneClient, kymaObjKey,
```
Comment (adityabhatia) [inline pkg/deploy/chart_manager.go:89]: can we store the `rest.Config` also so the cache along with `client.Client`, since they belong together and we do not have to do a lookup for the secret everytime

File: skr-webhook/templates/deployment.yaml [hunk 4]
```diff
@@ -30,7 +30,7 @@ spec:
     metadata:
       labels:
         app: skr-webhook
-        operator.kyma-project.io/pod-restart-trigger: "{{.Values.triggerLabel}}"
+        operator.kyma-project.io/pod-restart-trigger: {{uuidv4}}
```
Comment (adityabhatia) [inline skr-webhook/templates/deployment.yaml:33]: `pod-restart-trigger ` this label we do not need anymore, as we discussed the parsed Manifest will be cached. Could you please remove the trigger label logic?

File: skr-webhook/templates/rbac.yaml [hunk 5]
```diff
@@ -11,20 +11,20 @@ metadata:
   namespace: {{.Release.Namespace}}
   name: kyma-reader
 rules:
-- apiGroups: ["operator.kyma-project.io"] # "" indicates the core API group
-  resources: ["*"]
-  verbs: ["get", "watch", "list"]
+  - apiGroups: [ "operator.kyma-project.io" ] # "" indicates the core API group
+    resources: [ "*" ]
+    verbs: [ "get", "watch", "list" ]
```
Comment (adityabhatia) [inline skr-webhook/templates/rbac.yaml:16]: this formatting was fine before

File: skr-webhook/templates/rbac.yaml [hunk 6]
```diff
@@ -11,20 +11,20 @@ metadata:
...
   namespace: {{.Release.Namespace}}
 subjects:
-# You can specify more than one "subject"
-- kind: ServiceAccount
-  name: {{.Release.Name}}-webhook-sa
-  namespace: {{.Release.Namespace}}
+  # You can specify more than one "subject"
+  - kind: ServiceAccount
+    name: {{.Release.Name}}-webhook-sa
+    namespace: {{.Release.Namespace}}
```
Comment (adityabhatia) [inline skr-webhook/templates/rbac.yaml:27]: this formatting was fine before

File: skr-webhook/values.yaml [hunk 7]
```diff
@@ -2,15 +2,13 @@
...
-image: 
+image:
   # -- Image repository
-  repository: "eu.gcr.io/kyma-project"
+  repository: "europe-docker.pkg.dev/kyma-project/dev"
   # -- Image name
   name: "runtime-watcher-skr"
   # -- Image tag
-  tag: "latest"
+  tag: "PR-80"
```
Comment (adityabhatia) [inline skr-webhook/values.yaml:11]: please revert to latest

File: pkg/deploy/webhook_chart.go [hunk 8]
```diff
@@ -1,61 +1,122 @@
...
+	baseRes := &unstructured.Unstructured{}
+	baseRes.SetLabels(map[string]string{
+		modulelabels.CacheKey: kymaObjKey.String(),
+	})
+	baseRes.SetNamespace(metav1.NamespaceDefault)
+	baseRes.SetName(staticWatcherConfigName)
+	return baseRes
+}
+
+func CheckWebhookCABundleConsistency(ctx context.Context, skrClient client.Client, kymaObjKey client.ObjectKey) error {
```
Comment (adityabhatia) [inline pkg/deploy/webhook_chart.go:99]: As far as I see `CheckWebhookCABundleConsistency` is only used in tests. Then why is it part of this file?

File: pkg/deploy/webhook_chart.go [hunk 9]
```diff
@@ -1,61 +1,122 @@
...
-	DeploymentNameTpl      = "%s-webhook"
-	triggerLabelTimeFormat = "200601021504050700"
+	customConfigKey         = "modules"
+	IstioSytemNs            = "istio-system"
+	IngressServiceName      = "istio-ingressgateway"
+	DeploymentNameTpl       = "%s-webhook"
+	releaseNameTpl          = "%s-%s-skr"
+	staticWatcherConfigName = "static-watcher-config-name"
+	WebhookConfigNameTpl    = "%s-webhook"
+	SecretNameTpl           = "%s-webhook-tls" //nolint:gosec
```
Comment (adityabhatia) [inline pkg/deploy/webhook_chart.go:32]: remove lint, by renaming `SecretNameTpl`

Comment (adityabhatia) [inline pkg/deploy/webhook_chart.go:32]: `%s-webhook` is used 3 times. Could we re-use?

File: controllers/withwatcher/kyma_controller_test.go [hunk 10]
```diff
@@ -21,90 +27,212 @@ import (
...
+	ErrWebhookNamePartsNumberMismatch  = errors.New("webhook name dot separated parts number mismatch")
+	ErrManagedByLabelNotFound          = errors.New("managed-by label not found")
+	ErrModuleNameMismatch              = errors.New("module name mismatch")
+	ErrSvcPathMismatch                 = errors.New("service path mismatch")
+	ErrWatchLabelsMismatch             = errors.New("watch labels mismatch")
+	ErrStatusSubResourcesMismatch      = errors.New("status sub-resources mismatch")
+	ErrSpecSubResourcesMismatch        = errors.New("spec sub-resources mismatch")
+)
 
+var _ = Describe("Kyma with multiple module CRs in remote sync mode", Ordered, Serial, func() {
```
Comment (adityabhatia) [inline controllers/withwatcher/kyma_controller_test.go:49]: Ordered - I thought we decided to get rid of them?
Is setting Serial by mistake?

File: controllers/withwatcher/watcher_controller_test.go [hunk 11]
```diff
@@ -48,37 +48,26 @@ func allCRsDeleted() func(customIstioClient *istio.Client) {
...
 		}
 		// verify
-		Eventually(isCrDeletionFinished(), Timeout, Interval).Should(BeTrue())
+		Eventually(isWatcherCrDeletionFinished(), Timeout, Interval).Should(BeTrue())
 		Eventually(isVsRemoved(suiteCtx, customIstioClient)).Should(BeTrue())
 	}
 }
 
-var _ = Describe("Watcher CR scenarios", Ordered, func() {
+var _ = Describe("Watcher CR scenarios", Ordered, Serial, func() {
```
Comment (adityabhatia) [inline controllers/withwatcher/watcher_controller_test.go:56]: Ordered - I thought we decided to get rid of them?
Is setting Serial by mistake?

File: controllers/withwatcher/kyma_controller_test.go [hunk 12]
```diff
@@ -21,90 +27,212 @@ import (
...
+			Namespace: metav1.NamespaceDefault,
+			Name:      deploy.ResolveSKRChartResourceName(deploy.WebhookConfigNameTpl, kymaObjKey),
+		}, webhookConfig), Timeout, Interval).Should(Succeed())
+		Expect(isWebhookConfigured(watcherCrForKyma, webhookConfig, kyma.Name)).To(Succeed())
+	})
+
+	It("kyma reconciler re-installs watcher helm chart when webhook CA bundle is not consistent", func() {
+		By("updating webhook config with corrupt CA bundle data")
+		Expect(webhookConfig.Webhooks).NotTo(BeEmpty())
+		webhookConfig.Webhooks[0].ClientConfig.CABundle = []byte(dummyCaBundleData)
```
Comment (adityabhatia) [inline controllers/withwatcher/kyma_controller_test.go:77]: are you updating this `dummyCaBundleData`. As far as I see it's only set on the variable and not in the testenv cluster?

File: controllers/withwatcher/kyma_controller_test.go [hunk 13]
```diff
@@ -21,90 +27,212 @@ import (
...
+		By("updating kyma channel to trigger its reconciliation")
+		kyma.Spec.Channel = v1alpha1.ChannelRapid
+		Expect(controlPlaneClient.Update(suiteCtx, kyma)).To(Succeed())
+		By("waiting for the kyma update event to be processed")
+		time.Sleep(2 * time.Second)
```
Comment (adityabhatia) [inline controllers/withwatcher/kyma_controller_test.go:82]: this should not be required

File: controllers/withwatcher/kyma_controller_test.go [hunk 14]
```diff
@@ -21,90 +27,212 @@ import (
...
+		watcherCrForKyma.Spec.LabelsToWatch[labelKey] = labelValue
+		Expect(controlPlaneClient.Update(suiteCtx, watcherCrForKyma)).To(Succeed())
+		By("waiting for watcher CR labelsToWatch to be updated")
+		Eventually(isWatcherCrLabelUpdated(client.ObjectKeyFromObject(watcherCrForKyma),
+			labelKey, labelValue), Timeout, Interval).Should(BeTrue())
+		By("updating kyma channel to trigger its reconciliation")
+		kyma.Spec.Channel = v1alpha1.ChannelFast
+		Expect(controlPlaneClient.Update(suiteCtx, kyma)).To(Succeed())
+		By("waiting for the kyma update event to be processed")
+		time.Sleep(2 * time.Second)
```
Comment (adityabhatia) [inline controllers/withwatcher/kyma_controller_test.go:101]: could we please get rid of `time.Sleep`?

File: controllers/withwatcher/kyma_controller_test.go [hunk 15]
```diff
@@ -21,90 +27,212 @@ import (
...
+			Name:      deploy.ResolveSKRChartResourceName(deploy.WebhookConfigNameTpl, kymaObjKey),
+		}, webhookConfig)).To(Succeed())
+		Expect(isWebhookConfigured(watcherCrForKyma, webhookConfig, kyma.Name)).To(Succeed())
+
+		caBundleValueBeforeUpdate := webhookConfig.Webhooks[0].ClientConfig.CABundle
+		By("updating kyma channel to trigger its reconciliation")
+		kyma.Spec.Channel = v1alpha1.ChannelRapid
+		Expect(controlPlaneClient.Update(suiteCtx, kyma)).To(Succeed())
+		By("waiting for the kyma update event to be processed")
+		time.Sleep(2 * time.Second)
```
Comment (adityabhatia) [inline controllers/withwatcher/kyma_controller_test.go:116]: could we remove `time.Sleep`?

File: pkg/deploy/chart_manager.go [hunk 16]
```diff
@@ -6,26 +6,28 @@ import (
...
+const caCertificateSecretKey = "ca.crt"
+
 var ErrExpectedExactlyOneSKRConfig = errors.New("expected exactly one SKR config")
 
 type SKRWebhookChartManager interface {
-	InstallWebhookChart(ctx context.Context, kyma *v1alpha1.Kyma,
-		remoteClientCache *remote.ClientCache, kcpClient client.Client) (bool, error)
-	RemoveWebhookChart(ctx context.Context, kyma *v1alpha1.Kyma,
-		syncCtx *remote.KymaSynchronizationContext) error
+	InstallWebhookChart(ctx context.Context, syncContext *remote.KymaSynchronizationContext) (bool, error)
```
Comment (adityabhatia) [inline pkg/deploy/chart_manager.go:29]: I see that InstallWebhookChart is always called both during Processing and Ready States. I some webhook resource is missing during `Ready` State, it force applies. Could you write a test for this use case? I have my doubts it works..

File: controllers/kyma_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakobmoellersap) [reply controllers/kyma_controller.go:214]: Just because you use ctx during controller start does not mean the values set during the control-loop propagate upstream or cross boundaries.

If you call `context.WithValue(ctx, key, value)` in `Reconcile`, it will only pass the object in the context used within one reconciliation, thus it is safe to use even with a single key, there will be no references after Reconcile is over, and it will also be GCed normally.

File: pkg/deploy/chart_manager.go (hunk 16 above)
Comment (jakobmoellersap) [reply pkg/deploy/chart_manager.go:29]: Looking at the comment above: Force apply should never be a regular case imo and we should only use 3-way merge. Is that happening / possible right now?

File: pkg/deploy/chart_manager.go [hunk 17]
```diff
@@ -6,26 +6,28 @@ import (
...
+	RemoveWebhookChart(ctx context.Context, syncContext *remote.KymaSynchronizationContext) error
 }
 
 type DisabledSKRWebhookChartManager struct{}
```
Comment (jakobmoellersap) [inline pkg/deploy/chart_manager.go:33]: Imo this implementation should change to not use a `Disabled` struct. It should not be created in the first place. This kind of design reminds me a lot of Java interface implementations...

File: controllers/kyma_controller.go [hunk 18]
```diff
@@ -286,15 +289,13 @@ func (r *KymaReconciler) HandleProcessingState(ctx context.Context, kyma *v1alph
...
 	logger := log.FromContext(ctx)
 
-	if kyma.Spec.Sync.Enabled {
-		syncContext, err := remote.InitializeKymaSynchronizationContext(ctx, kyma, r.Client, r.RemoteClientCache)
-		if err != nil {
-			return false, fmt.Errorf("remote sync initialization failed: %w", err)
-		}
-		if err = r.RemoveWebhookChart(ctx, kyma, syncContext); err != nil {
+	if kyma.Spec.Sync.Enabled && syncContext != nil {
+		if err := r.RemoveWebhookChart(ctx, syncContext); err != nil {
```
Comment (jakobmoellersap) [inline controllers/kyma_controller.go:298]: why has remove web hook chart no dedication switch between recoverable and non-recoverable errors? e.g. a chart installation is not something that can necessarily be recovered simply by waiting (and returning true, nil as you do)

File: pkg/deploy/webhook_chart.go [hunk 19]
```diff
@@ -1,61 +1,122 @@
...
-	ModeUninstall          = Mode("uninstall")
-	customConfigKey        = "modules"
-	ReleaseName            = "skr"
-	IstioSytemNs           = "istio-system"
-	IngressServiceName     = "istio-ingressgateway"
-	DeploymentNameTpl      = "%s-webhook"
-	triggerLabelTimeFormat = "200601021504050700"
+	customConfigKey         = "modules"
+	IstioSytemNs            = "istio-system"
+	IngressServiceName      = "istio-ingressgateway"
```
Comment (jakobmoellersap) [inline pkg/deploy/webhook_chart.go:27]: I still think this should be configurable in the controller via flag, same as the namespace. An Istio-installation can look different than this and it heavily opinionated

File: skr-webhook/templates/rbac.yaml [hunk 20]
```diff
@@ -11,20 +11,20 @@ metadata:
   namespace: {{.Release.Namespace}}
   name: kyma-reader
 rules:
-- apiGroups: ["operator.kyma-project.io"] # "" indicates the core API group
-  resources: ["*"]
-  verbs: ["get", "watch", "list"]
+  - apiGroups: [ "operator.kyma-project.io" ] # "" indicates the core API group
+    resources: [ "*" ]
```
Comment (jakobmoellersap) [inline skr-webhook/templates/rbac.yaml:15]: The star import is not strict enough, why is this here?

File: skr-webhook/values.yaml [hunk 21]
```diff
@@ -2,15 +2,13 @@
 sidecar: true
 # -- Replicas
 replicas: 1
-# -- Trigger pod restart
-triggerLabel: ""
-image: 
+image:
   # -- Image repository
-  repository: "eu.gcr.io/kyma-project"
+  repository: "europe-docker.pkg.dev/kyma-project/dev"
```
Comment (jakobmoellersap) [inline skr-webhook/values.yaml:7]: please also revert this to the proper registry

File: controllers/watcher_controller.go [hunk 22]
```diff
@@ -66,7 +66,7 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 
 	watcherObj := &v1alpha1.Watcher{}
 	if err := r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, watcherObj); err != nil {
-		logger.Error(err, "Failed to get reconciliation object")
+		logger.V(1).Info("Failed to get reconciliation object")
```
Comment (jakobmoellersap) [inline controllers/watcher_controller.go:69]: Why is V1 hardcoded here?

File: pkg/deploy/chart_manager.go [hunk 23]
```diff
@@ -74,81 +76,97 @@ type SkrChartConfig struct {
...
-	if err != nil {
-		return true, err
+	if !installed {
+		return true, ErrSKRWebhookNotInstalled
 	}
-	kyma.UpdateCondition(v1alpha1.ConditionReasonSKRWebhookIsReady, metav1.ConditionTrue)
+	syncContext.ControlPlaneKyma.UpdateCondition(v1alpha1.ConditionReasonSKRWebhookIsReady, metav1.ConditionTrue)
+	logger.Info("successfully installed webhook chart",
+		"release-name", skrWatcherInstallInfo.ChartInfo.ReleaseName)
+	logger.V(1).Info("following modules were installed",
```
Comment (jakobmoellersap) [inline pkg/deploy/chart_manager.go:110]: again, why is V1 hardcoded?

File: skr-webhook/templates/deployment.yaml (hunk 4 above)
Comment (khlifi411) [reply skr-webhook/templates/deployment.yaml:33]: I think we still need it. when the set flags' values change, only the `webhook` config and secret will change and the pods will not be restarted. unless this label value changes.

File: controllers/withwatcher/kyma_controller_test.go (hunk 15 above)
Comment (khlifi411) [reply controllers/withwatcher/kyma_controller_test.go:121]: I am not sure we can. I tried different post checks without success. what do you suggest?

File: controllers/kyma_controller.go (hunk 18 above)
Comment (jakobmoellersap) [reply controllers/kyma_controller.go]: agree on this, no way for us here.

File: controllers/kyma_controller.go [hunk 24]
```diff
@@ -231,48 +232,20 @@ func (r *KymaReconciler) HandleInitialState(ctx context.Context, kyma *v1alpha1.
...
-	modules, err = r.GenerateModulesFromTemplate(ctx, kyma)
-	if err != nil {
-		return r.UpdateStatusWithEventFromErr(ctx, kyma, v1alpha1.StateError,
-			fmt.Errorf("error while fetching modules during processing: %w", err))
-	}
-
-	runner := sync.New(r)
-
-	statusUpdateRequiredFromModuleSync, err := runner.Sync(ctx, kyma, modules)
+	statusUpdateRequiredFromModulesSync, err := r.syncModules(ctx, kyma)
```
Comment (adityabhatia) [inline controllers/kyma_controller.go:235]: any particular reason to move this out?

File: pkg/deploy/template_chart_manager.go [hunk 25]
```diff
@@ -0,0 +1,196 @@
...
+		oldResource.SetGroupVersionKind(resource.GroupVersionKind())
+		err := syncContext.RuntimeClient.Get(ctx, resourceObjKey, oldResource)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return true, fmt.Errorf("failed to get webhook %s: %w", resource.GetKind(), err)
+		}
+		if apierrors.IsNotFound(err) {
+			if err := syncContext.RuntimeClient.Create(ctx, resource, defaultFieldOwner); err != nil {
+				return true, fmt.Errorf("failed to create webhook %s: %w", resource.GetKind(), err)
+			}
+		}
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:126]: you can directly use `syncContext.RuntimeClient.Create`. If the resource exists it will anyways result in a conflict and report an error

File: controllers/kyma_controller.go (hunk 24 above)
Comment (khlifi411) [reply controllers/kyma_controller.go:235]: I moved this part to a separate method because of the cyclomatic complexity linting error. see https://status.build.kyma-project.io/view/gs/kyma-prow-logs/pr-logs/pull/kyma-project_lifecycle-manager/313/pull-lifecycle-mgr-lint/1613491447321858048

File: pkg/deploy/template_chart_manager.go [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+	for _, resource := range resources {
+		resourceObjKey := client.ObjectKeyFromObject(resource)
+		oldResource := &unstructured.Unstructured{}
+		oldResource.SetGroupVersionKind(resource.GroupVersionKind())
+		err := syncContext.RuntimeClient.Get(ctx, resourceObjKey, oldResource)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return true, fmt.Errorf("failed to get webhook %s: %w", resource.GetKind(), err)
+		}
+		if apierrors.IsNotFound(err) {
+			if err := syncContext.RuntimeClient.Create(ctx, resource, defaultFieldOwner); err != nil {
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:123]: why are you passing a defaultFieldOwner? 
You're using client-side apply here.

File: pkg/deploy/template_chart_manager.go [hunk 27]
```diff
@@ -0,0 +1,196 @@
...
+				return true, fmt.Errorf("failed to create webhook %s: %w", resource.GetKind(), err)
+			}
+		}
+		// completely replace old resource with new resource
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:127]: considering that the resource was freshly created, why is it being updated right after. Somehow this doesn't make sense.

File: pkg/deploy/template_chart_manager.go [hunk 28]
```diff
@@ -0,0 +1,196 @@
...
+		oldResource := &unstructured.Unstructured{}
+		oldResource.SetGroupVersionKind(resource.GroupVersionKind())
+		err := syncContext.RuntimeClient.Get(ctx, resourceObjKey, oldResource)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get webhook %s: %w", resource.GetKind(), err)
+		}
+		if err == nil {
+			if err := syncContext.RuntimeClient.Delete(ctx, resource); err != nil {
+				return fmt.Errorf("failed to delete webhook %s: %w", resource.GetKind(), err)
+			}
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:171]: you can directly delete the resource without calling `Get()`

File: pkg/deploy/template_chart_manager.go [hunk 29]
```diff
@@ -0,0 +1,196 @@
...
+
+func NewSKRWebhookTemplateChartManager(kcpRestConfig *rest.Config, config *SkrChartManagerConfig,
+) (*SKRWebhookTemplateChartManager, error) {
+	resolvedKcpAddr, err := resolveKcpAddr(kcpRestConfig, config)
+	if err != nil {
+		return nil, err
+	}
+	return &SKRWebhookTemplateChartManager{
+		config:  config,
+		cache:   &sync.Map{},
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:57]: can we rename `cache` to something like `configCache`

File: pkg/deploy/template_chart_manager.go [hunk 30]
```diff
@@ -0,0 +1,196 @@
...
+		Values:      chartArgValues,
+	})
+}
+
+func (m *SKRWebhookTemplateChartManager) cachedConfig(kymaObjKey client.ObjectKey,
+	chartArgValues map[string]interface{},
+) bool {
+	cachedArgValues, ok := m.cache.Load(kymaObjKey)
+	if !ok {
+		m.cache.Store(kymaObjKey, chartArgValues)
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:83]: why are you string chartConfig by kyma name? AFAIK, all kymas will have the same config. Shouldn't we simplify this just store config?

**Review by adityabhatia — CHANGES_REQUESTED** [review]
Looks good overall, apart from the comments, inline comments are missing.

File: Makefile [hunk 31]
```diff
@@ -185,6 +185,10 @@ lister-gen: ## Download lister-gen
 fmt: ## Run go fmt against code.
 	go fmt ./...
 
+.PHONY: resolve
+resolve: ## Run go mod tidy against code.
```
Comment (jakobmoellersap) [inline Makefile:189]: could we try to get rid of these resolves down the line? I would like to avoid having to pin golang versions like we had to for the old reconciler

File: go.mod [hunk 32]
```diff
@@ -7,11 +7,11 @@ require (
 	github.com/gardener/component-spec/bindings-go v0.0.70
 	github.com/go-logr/logr v1.2.3
 	github.com/go-logr/zapr v1.2.3
-	github.com/kyma-project/module-manager v0.0.0-20221207164018-ddf69229acb6
-	github.com/kyma-project/module-manager/operator v0.0.0-20221125150159-5e84c4855249
+	github.com/kyma-project/module-manager v0.0.0-20221214222100-348f6d59ed0b
 	github.com/kyma-project/runtime-watcher/listener v0.0.0-20221129135108-60e6093eba9d
 	github.com/onsi/ginkgo/v2 v2.5.0
 	github.com/onsi/gomega v1.24.1
+	github.com/slok/go-helm-template v0.3.0
```
Comment (jakobmoellersap) [inline go.mod:14]: this still includes the helm libraries, did we not agree to get rid of the entire chart?

**Review by jakobmoellersap — COMMENTED** [review]
just wondering in general why we don't entirely remove helm from the dependency list. If you want to do that in a new step, also fine by me

File: go.mod (hunk 32 above)
Comment (khlifi411) [reply go.mod:14]: Yes. But we will do it in a separate ticket: #346

File: pkg/deploy/template_chart_manager.go [hunk 33]
```diff
@@ -0,0 +1,146 @@
...
+		resourceObjKey := client.ObjectKeyFromObject(resource)
+		oldResource := &unstructured.Unstructured{}
+		oldResource.SetGroupVersionKind(resource.GroupVersionKind())
+		if err := skrClient.Get(ctx, resourceObjKey, oldResource); err != nil {
+			return fmt.Errorf("failed to get webhook %s: %w", resource.GetKind(), err)
+		}
+		resource.SetResourceVersion(oldResource.GetResourceVersion())
+		if err := skrClient.Update(ctx, resource); err != nil {
+			return fmt.Errorf("failed to replace webhook %s: %w", resource.GetKind(), err)
+		}
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:142]: maybe you can simply use patch here.

File: pkg/deploy/template_chart_manager.go [hunk 34]
```diff
@@ -0,0 +1,146 @@
...
+		return fmt.Errorf("failed to create webhook %s: %w", resource.GetKind(), err)
+	}
+	if resourceAlreadyExists {
+		resourceObjKey := client.ObjectKeyFromObject(resource)
+		oldResource := &unstructured.Unstructured{}
+		oldResource.SetGroupVersionKind(resource.GroupVersionKind())
+		if err := skrClient.Get(ctx, resourceObjKey, oldResource); err != nil {
+			return fmt.Errorf("failed to get webhook %s: %w", resource.GetKind(), err)
+		}
+		resource.SetResourceVersion(oldResource.GetResourceVersion())
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:139]: why setting this explicitly?

Comment (khlifi411) [reply pkg/deploy/template_chart_manager.go:139]: In CSA, without the resource version field set, the `Update()` call fails. As agreed already in the daily, this implementation will be replaced with the SSA implementation.

File: pkg/deploy/template_chart_manager.go (hunk 33 above)
Comment (khlifi411) [reply pkg/deploy/template_chart_manager.go:142]: As already discussed in the daily, SSA will replace this implementation.

**Review by jakobmoellersap — COMMENTED** [review]
LGTM, good stuff! waiting for other reviews

File: pkg/deploy/template_chart_manager.go [hunk 35]
```diff
@@ -0,0 +1,120 @@
...
+	manifest, err := renderChartToRawManifest(ctx, kymaObjKey, m.config.WebhookChartPath, chartArgValues)
+	if err != nil {
+		return true, err
+	}
+	resources, err := getRawManifestUnstructuredResources(manifest)
+	if err != nil {
+		return true, err
+	}
+	for _, resource := range resources {
+		err := syncContext.RuntimeClient.Patch(ctx, resource, client.Apply, client.ForceOwnership, skrChartFieldOwner)
```
Comment (jakobmoellersap) [inline pkg/deploy/template_chart_manager.go:70]: would you kindly update this to a concurrent implementation using ErrGroup? otherwise we will have to wait for each resource patch to complete, which develops bottlenecking here

File: pkg/deploy/template_chart_manager.go [hunk 36]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	resources, err := getRawManifestUnstructuredResources(manifest)
+	if err != nil {
+		return err
+	}
+	for _, resource := range resources {
+		if err := syncContext.RuntimeClient.Delete(ctx, resource); err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to delete webhook %s: %w", resource.GetKind(), err)
+		}
+	}
```
Comment (adityabhatia) [inline pkg/deploy/template_chart_manager.go:104]: does it make sense to implement similar errGrp here, so that deletion of following resources are not blocked by one failing?

**Review by adityabhatia — APPROVED** [review]
